consul/autopilot.go

```diff
@@ -0,0 +1,286 @@
+package consul
+
+import (
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/hashicorp/consul/consul/agent"
+	"github.com/hashicorp/consul/consul/structs"
+	"github.com/hashicorp/raft"
+	"github.com/hashicorp/serf/serf"
+)
+
+func (s *Server) startAutopilot() {
+	s.autopilotShutdownCh = make(chan struct{})
+
+	go s.serverHealthLoop()
+	go s.removeDeadLoop()
+}
+
+func (s *Server) stopAutopilot() {
+	close(s.autopilotShutdownCh)
+}
+
+// serverHealthLoop monitors the health of the servers in the cluster
+func (s *Server) serverHealthLoop() {
+	// Monitor server health until shutdown
+	ticker := time.NewTicker(s.config.ServerHealthInterval)
+	for {
+		select {
+		case <-s.autopilotShutdownCh:
+			ticker.Stop()
+			return
+		case <-ticker.C:
+			serverHealths := make(map[string]*structs.ServerHealth)
+
+			state := s.fsm.State()
+			_, autopilotConf, err := state.AutopilotConfig()
+			if err != nil {
+				s.logger.Printf("[ERR] consul: error retrieving autopilot config: %s", err)
+			}
+
+			// Build an updated map of server healths
+			for _, member := range s.LANMembers() {
+				if member.Status == serf.StatusLeft {
+					continue
+				}
+
+				valid, parts := agent.IsConsulServer(member)
+				if valid {
+					health := s.queryServerHealth(member, parts, autopilotConf)
+					serverHealths[parts.Addr.String()] = health
+				}
+			}
+
+			s.autopilotLock.Lock()
+			s.autopilotHealth = serverHealths
+			s.autopilotLock.Unlock()
+
+			if err := s.promoteNonVoters(autopilotConf); err != nil {
+				s.logger.Printf("[ERR] consul: error checking for non-voters to promote: %s", err)
+			}
+		}
+	}
+}
+
+// removeDeadLoop checks for dead servers periodically, or when receiving on autopilotRemoveDeadCh
+func (s *Server) removeDeadLoop() {
+	ticker := time.NewTicker(s.config.RemoveDeadInterval)
+	for {
+		select {
+		case <-s.autopilotShutdownCh:
+			ticker.Stop()
+			return
+		case <-ticker.C:
+			if err := s.pruneDeadServers(); err != nil {
+				s.logger.Printf("[ERR] consul: error checking for dead servers to remove: %s", err)
+			}
+		case <-s.autopilotRemoveDeadCh:
+			if err := s.pruneDeadServers(); err != nil {
+				s.logger.Printf("[ERR] consul: error checking for dead servers to remove: %s", err)
+			}
+		}
+	}
+}
+
+// pruneDeadServers removes up to numPeers/2 failed servers
+func (s *Server) pruneDeadServers() error {
+	state := s.fsm.State()
+	_, autopilotConf, err := state.AutopilotConfig()
+	if err != nil {
+		return err
+	}
+
+	// Look for dead servers to clean up
+	peers, err := s.numPeers()
+	if err != nil {
+		return err
+	}
+
+	removed := 0
+	if autopilotConf.CleanupDeadServers {
+		for _, member := range s.serfLAN.Members() {
+			// Exit early if we already removed the max amount of servers
+			if removed == peers/2 {
```

Comment:
I wonder if it's safer to check to see how many it would remove and don't remove any if this threshold will be hit. This will chip away at a large number over subsequent runs.